autoload/go/term.vim

```diff
@@ -30,49 +30,73 @@ function! go#term#newmode(bang, cmd, errorformat, mode) abort
-  " refer to state. See :help Partial for more information.
-  "
-  " Don't set an on_stderr, because it will be passed the same data as
-  " on_stdout. See https://github.com/neovim/neovim/issues/2836
-  let l:job = {
-        \ 'on_stdout': function('s:on_stdout', [], state),
-        \ 'on_exit' : function('s:on_exit', [], state),
-      \ }
-
-  let l:state.id = termopen(a:cmd, l:job)
-  let l:state.termwinid = win_getid(winnr())
-
-  execute l:cd . fnameescape(l:dir)
-
-  " resize new term if needed.
   let l:height = go#config#TermHeight()
   let l:width = go#config#TermWidth()
 
-  " Adjust the window width or height depending on whether it's a vertical or
-  " horizontal split.
-  if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
-    exe 'vertical resize ' . l:width
-  elseif mode =~ "split" || mode =~ "new"
-    exe 'resize ' . l:height
-  endif
+  " setup job for nvim
+  if has('nvim')
+    execute l:mode . ' __go_term__'
+    " explicitly bind callbacks to state so that within them, self will always
+    " refer to state. See :help Partial for more information.
+    "
+    " Don't set an on_stderr, because it will be passed the same data as
+    " on_stdout. See https://github.com/neovim/neovim/issues/2836
+    let l:job = {
+          \ 'on_stdout': function('s:on_stdout', [], state),
+          \ 'on_exit' : function('s:on_exit', [], state),
+        \ }
+    let l:state.id = termopen(a:cmd, l:job)
+    let l:state.termwinid = win_getid(winnr())
+    execute l:cd . fnameescape(l:dir)
+
+    " Adjust the window width or height depending on whether it's a vertical or
+    " horizontal split.
+    if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
+      exe 'vertical resize ' . l:width
+    elseif mode =~ "split" || mode =~ "new"
+      exe 'resize ' . l:height
+    endif
+    " we also need to resize the pty, so there you go...
+    call jobresize(l:state.id, l:width, l:height)
+
+  " setup term for vim8
+  elseif has('terminal')
+    let l:term = {
+          \ 'term_rows' : l:height,
+          \ 'term_cols' : l:width,
+          \ 'out_cb': function('s:out_cb', [], state),
+          \ 'exit_cb' : function('s:exit_cb', [], state),
+        \ }
+
+    if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
+      let l:term["vertical"] = 1
+    endif
 
-  " we also need to resize the pty, so there you go...
-  call jobresize(l:state.id, l:width, l:height)
+    let l:id = term_start(["/bin/sh", "-c", a:cmd], l:term)
+    let l:state.id = l:id
+    let l:state.termwinid = win_getid(bufwinnr(id))
+    execute l:cd . fnameescape(l:dir)
+  else 
+    return -1
+  endif
 
   call win_gotoid(l:state.winid)
-
   return l:state.id
 endfunction
 
+" out_cb continually concat's the self.stdout_buf on recv of stdout
+" and sets self.stdout to the new-lined split content in self.stdout_buf
+func! s:out_cb(channel, msg) dict abort
+  let self.stdout_buf = self.stdout_buf . a:msg
```

Comment:
@bhcleek 
I am not sure if the msg here is NL buffered or not...

Also this may not be necessary if the `exit_cb` callback can simply grab the contents of the terminal buffer id instead of accumulating a string buffer... my thinking is we are duplicating the data now.

Reply:
That seems like something that should be figured out before merging.